Suit sensor and crew monitoring

Macoron opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->
Add basic crew monitoring console and suit sensors.

It's a bit experiential, because suit sensor use device network to broadcast their state to crew monitoring consoles. It works like this:
1. Each two seconds sensor updates its state
2. State is packed into device network message and broadcasted over wireless net
3. Monitoring console receive message and save it in an internal buffer
4. If console hasn't received an state update from sensor for more than 10 seconds, it presumes that sensor was disabled or lost connection and remove it from buffer

As a side effect of last part: if you quickly change character jumpsuits,  crew monitoring console will display duplicates (though they will disappear quickly).

**Screenshots**
<!-- If applicable, add screenshots to showcase your PR. If your PR is a visual change, add
screenshots or it's liable to be closed by maintainers. -->
![image](https://user-images.githubusercontent.com/6161335/143502702-961ae147-2a6c-4c02-b042-0cc5d2e09ed3.png)


![image](https://user-images.githubusercontent.com/6161335/143502674-4c7408a8-34db-4f3e-bac3-643f19d4542a.png)


**Changelog**
<!--
Here you can fill out a changelog that will automatically be added to the game when your PR is merged
There are 4 icons for changelog entries: add, remove, tweak, fix. I trust you can figure out the rest.

You can put your name after the :cl: symbol to change the name that shows in the changelog (otherwise it takes your GitHub username)
Like so: :cl: PJB

Generally, only put things in changelogs that players actually care about. Stuff like "Refactored X system, no changes should be visible" shouldn't be on a changelog.

For writing actual entries, don't consider the entry type suffix (e.g. add) to be "part" of the sentence:
bad: - add: a new tool for engineers
good: - add: added a new tool for engineers
-->

:cl:
- add: Added suit sensors and crew monitoring console in code. They are not mapped, but you can print circuit board.
- add: Prisoners suit has locked suit sensor that always broadcasting their position.

moonheart08 commented:
Probably needs adjusted for [5422] should that be merged first

juliangiebel commented on Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:
```diff
@@ -0,0 +1,291 @@
+            return payload;
+        }
+
+        public SuitSensorStatus? PackageToSuitSensor(NetworkPayload payload)
```
It's called packet so you should probably name that method PacketToSuitSensor 😄

juliangiebel commented on Resources/Prototypes/Entities/Clothing/Uniforms/base.yml:
```diff
@@ -1,6 +1,17 @@
+  - type: DeviceNetworkComponent
+    deviceNetId: Wireless
+  - type: WirelessNetworkConnection
+    range: 500
```
Did you try out the range in game?

Macoron replied:
Do you mean out of range? Yes, worked fine

juliangiebel reviewed (approved):
Looks good to me 👍 
Usage off device networking is good 👌

PaulRitter reviewed (changes requested):
checking to see if my pending comment got eaten by the github crash

PaulRitter commented on Content.Client/Medical/CrewMonitoring/CrewMonitoringTableEntry.xaml.cs:
```diff
@@ -0,0 +1,16 @@
+using Robust.Client.AutoGenerated;
+using Robust.Client.UserInterface;
+using Robust.Client.UserInterface.CustomControls;
+using Robust.Client.UserInterface.XAML;
+
+namespace Content.Client.Medical.CrewMonitoring
+{
+    [GenerateTypedNameReferences]
+    public partial class CrewMonitoringTableEntry : Control
+    {
+        public CrewMonitoringTableEntry()
+        {
+            RobustXamlLoader.Load(this);
+        }
+    }
+}
```
i dont see this control used anywhere?

PaulRitter commented on Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:
```diff
@@ -0,0 +1,291 @@
+
+        private void OnMapInit(EntityUid uid, SuitSensorComponent component, MapInitEvent args)
+        {
+            // generate random mode
+            if (component.RandomMode)
+            {
+                //make the sensor mode favor higher levels, except coords.
+                var modesDist = new[]
+                {
+                    SuitSensorMode.SensorOff,
+                    SuitSensorMode.SensorBinary, SuitSensorMode.SensorBinary,
+                    SuitSensorMode.SensorVitals, SuitSensorMode.SensorVitals, SuitSensorMode.SensorVitals,
+                    SuitSensorMode.SensorCords, SuitSensorMode.SensorCords
+                };
+                component.Mode = _random.Pick(modesDist);
+            }
```
this should be done on componentinit, not mapinit imo

Macoron replied:
what the difference?

PaulRitter replied:
both work, but i'd still rather have it on compinit since its not maprelated.

compinit -> when the component gets intialized on entitycreation
mapinit -> when the map the entity is on gets intialized

Macoron replied:
> both work, but i'd still rather have it on compinit since its not maprelated.

But why? Right now if you save and load map - suit sensor will save its mode. But if you move it to init, it will be rewriting its state with each reload.

PaulRitter replied:
afaict mapinit should still run, even if you load a map

PaulRitter replied:
the idea is to have this run everytime a component is intialized, not everytime the map the components entity is on get initialized, since the first makes more sense

PaulRitter commented on Content.Server/Medical/SuitSensors/SuitSensorComponent.cs:
```diff
@@ -0,0 +1,60 @@
+        [DataField("updateRate")]
+        public float UpdateRate = 2f;
+
+        /// <summary>
+        ///     Current user that wears suit sensor. Null if nobody wearing it.
+        /// </summary>
+        public EntityUid? User = null;
```
pretty sure you can just get the currect wearer by getting the clothings parent (Transform.ParentUid)

Macoron replied:
I don't think its a good idea. Sounds like a dirty hack.
Equipped and Unequipped events are much more robust solution. It also allows to check that sensor was actually worn in a valid inventory slot.

PaulRitter commented on Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:
```diff
@@ -0,0 +1,291 @@
+            {
+                CreateVerb(uid, component, args.User.Uid, SuitSensorMode.SensorOff),
+                CreateVerb(uid, component, args.User.Uid, SuitSensorMode.SensorBinary),
+                CreateVerb(uid, component, args.User.Uid,SuitSensorMode.SensorVitals),
```
```suggestion
                CreateVerb(uid, component, args.User.Uid, SuitSensorMode.SensorVitals),
```

PaulRitter commented on Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:
```diff
@@ -0,0 +1,291 @@
+                    userJob = card.JobTitle;
+            }
+
+            // get health mob state
+            if (!EntityManager.TryGetComponent(sensor.User.Value, out MobStateComponent? mobState))
+                return null;
+            var isAlive = mobState.IsAlive();
+
+            // get mob total damage
+            if (!EntityManager.TryGetComponent(sensor.User.Value, out DamageableComponent? damageable))
+                return null;
+            var totalDamage = damageable.TotalDamage.Int();
+
+            // finally, form suit sensor status
+            var status = new SuitSensorStatus(userName, userJob);
+            switch (sensor.Mode)
+            {
+                case SuitSensorMode.SensorBinary:
+                    status.IsAlive = isAlive;
+                    break;
+                case SuitSensorMode.SensorVitals:
+                    status.IsAlive = isAlive;
+                    status.TotalDamage = totalDamage;
+                    break;
+                case SuitSensorMode.SensorCords:
+                    status.IsAlive = isAlive;
+                    status.TotalDamage = totalDamage;
+                    status.Coordinates = transform.MapPosition;
+                    break;
+            }
```
```suggestion
            // finally, form suit sensor status
            var status = new SuitSensorStatus(userName, userJob);
            switch (sensor.Mode)
            {
                case SuitSensorMode.SensorCords:
                    status.Coordinates = transform.MapPosition;
                case SuitSensorMode.SensorVitals:
                    // get mob total damage
                    if (EntityManager.TryGetComponent(sensor.User.Value, out DamageableComponent? damageable))
                        status.TotalDamage = damageable.TotalDamage.Int();
                case SuitSensorMode.SensorBinary:
                    // get health mob state
                    if (EntityManager.TryGetComponent(sensor.User.Value, out MobStateComponent? mobState))
                        status.IsAlive = mobState.IsAlive();
                    break;
            }
```
webedited together, check formatting after you commit this one or just code it yourself

Macoron replied:
I've checked in IDE. Code above isn't valid C# code. You can't "fallthrough" case statements in C#. Here is an example which wouldn't compily:
```
private enum X { A, B }

private void Boo()
{
    int i = 0;
    var e = X.A;

    switch (e)
    {
        case X.A:
            i++;
        case X.B:
            i--;
            break;
    }
}
```

PaulRitter replied:
my bad, you need to write `goto case X;` to create a fallthrough. implicit fallthroughs are disabled in c#, for good reason i might add, so you need to explicitely define them.

that being said, i forgot to mention why i actually wanted this change: i dont like simply a mobstatecomp/damageablecomp not existing killing the entire state from being sent.

Macoron replied:
I've added default values to current code, if components don't exist.

PaulRitter reviewed (changes requested):
it did

Macoron commented:
Addressed reviews and fixed merge conflict.

> Probably needs adjusted for #5422 should that be merged first

Tbh, I'm not sure, how to adjust this for multistation setup. Right now any suit sensor will be shown in crew monitoring console,  if they are in 500 tiles radius. The mob name and job are directly taken from IdCardSlot. Should I filter mobs somehow by grid id? Or maybe it would be better to use different wireless network channels?

If someone has an idea what should be multistation crew monitoring behavior, I can create an issue, after this PR will be merged.

moonheart08 commented:
> Addressed reviews and fixed merge conflict.
> 
> > Probably needs adjusted for #5422 should that be merged first
> 
> Tbh, I'm not sure, how to adjust this for multistation setup. Right now any suit sensor will be shown in crew monitoring console, if they are in 500 tiles radius. The mob name and job are directly taken from IdCardSlot. Should I filter mobs somehow by grid id? Or maybe it would be better to use different wireless network channels?
> 
> If someone has an idea what should be multistation crew monitoring behavior, I can create an issue, after this PR will be merged.

What could be done is simply including the StationId the sensor was originally registered on in the message, and only showing it in the console on the corresponding station. You can check what station a  grid is part of by grabbing the id from the grid's StationComponent

Macoron commented:
> > Addressed reviews and fixed merge conflict.
> > > Probably needs adjusted for #5422 should that be merged first
> > 
> > 
> > Tbh, I'm not sure, how to adjust this for multistation setup. Right now any suit sensor will be shown in crew monitoring console, if they are in 500 tiles radius. The mob name and job are directly taken from IdCardSlot. Should I filter mobs somehow by grid id? Or maybe it would be better to use different wireless network channels?
> > If someone has an idea what should be multistation crew monitoring behavior, I can create an issue, after this PR will be merged.
> 
> What could be done is simply including the StationId the sensor was originally registered on in the message, and only showing it in the console on the corresponding station. You can check what station a grid is part of by grabbing the id from the grid's StationComponent

I've discussed this issue in Discord. This would probably solved by new encryption keys mechanics for different stations in a separate PR.

PaulRitter commented:
> > > Addressed reviews and fixed merge conflict.
> > > > Probably needs adjusted for #5422 should that be merged first
> > > 
> > > 
> > > Tbh, I'm not sure, how to adjust this for multistation setup. Right now any suit sensor will be shown in crew monitoring console, if they are in 500 tiles radius. The mob name and job are directly taken from IdCardSlot. Should I filter mobs somehow by grid id? Or maybe it would be better to use different wireless network channels?
> > > If someone has an idea what should be multistation crew monitoring behavior, I can create an issue, after this PR will be merged.
> > 
> > 
> > What could be done is simply including the StationId the sensor was originally registered on in the message, and only showing it in the console on the corresponding station. You can check what station a grid is part of by grabbing the id from the grid's StationComponent
> 
> I've discussed this issue in Discord. This would probably solved by new encryption keys mechanics for different stations in a separate PR.

if you want to implement that, feel free to. I'd be happy with you doing what moony said, considering the encryption keys out of scope. Your call.

Macoron commented:
> > > > Addressed reviews and fixed merge conflict.
> > > > > Probably needs adjusted for #5422 should that be merged first
> > > > 
> > > > 
> > > > Tbh, I'm not sure, how to adjust this for multistation setup. Right now any suit sensor will be shown in crew monitoring console, if they are in 500 tiles radius. The mob name and job are directly taken from IdCardSlot. Should I filter mobs somehow by grid id? Or maybe it would be better to use different wireless network channels?
> > > > If someone has an idea what should be multistation crew monitoring behavior, I can create an issue, after this PR will be merged.
> > > 
> > > 
> > > What could be done is simply including the StationId the sensor was originally registered on in the message, and only showing it in the console on the corresponding station. You can check what station a grid is part of by grabbing the id from the grid's StationComponent
> > 
> > 
> > I've discussed this issue in Discord. This would probably solved by new encryption keys mechanics for different stations in a separate PR.
> 
> if you want to implement that, feel free to. I'd be happy with you doing what moony said, considering the encryption keys out of scope. Your call.

Tbh, I don't see any issue with current implementation. Not a big deal if players can see other players from another station in 500 tiles radius. Sounds more like a feature to me.

So I would like to leave this PR as is and address it in other PR.

mirrorcult commented:
leaving this as-is and encryption keys for another PR sounds fine to me

moonheart08 commented:
> leaving this as-is and encryption keys for another PR sounds fine to me

Sounds fine to me.

Zumorica commented:
Conflicts

Macoron commented:
Oh, thanks Sloth. I was too busy on my work, didn't have time to fix merge conflict by myself. Thank you for your help.
